### PR TITLE
[#1510] feat(trino-connector): Support partition, distribution and sort order of Hive table created by Trino

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

We can create a Hive table with partitioning, distribution, and sorting ordered by Trino.

### Why are the changes needed?

It's a crucial feature of the Trino connector. 

Fix: #1510 

### Does this PR introduce _any_ user-facing change?

User can create a hive table in Trino by:
```
create table t10 (id int, name varchar) with (partitioned_by = ARRAY['name'], bucketed_by = ARRAY['id'], bucket_count = 50);
```
Or  create a Hive table and loaded by Trino like 
```
trino:db2> show create table t11;
                                   Create Table
----------------------------------------------------------------------------------
 CREATE TABLE "test.hive_catalog".db2.t11 (
    id integer,
    name varchar(65535)
 )
 COMMENT ''
 WITH (
    bucket_count = 50,
    bucketed_by = ARRAY['id'],
    input_format = 'org.apache.hadoop.mapred.TextInputFormat',
    location = 'hdfs://localhost:9000/user/hive/warehouse/db2.db/t11',
    output_format = 'org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat',
    partitioned_by = ARRAY['name'],
    serde_lib = 'org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe',
    serde_name = 't11',
    table_type = 'MANAGED_TABLE'
 )
(1 row)
```

### How was this patch tested?

Add some IT

### CODE REVIEW

**Comment by yuqi1129** [comment]
I will convert this to a draft as it depends on item #1455

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/trino/TrinoConnectorIT.java [hunk 1]
```diff
@@ -388,6 +397,14 @@ void testHiveTableCreatedByTrino() {
     Assertions.assertEquals(
         "hdfs://localhost:9000/user/hive/warehouse/hive_schema.db/hive_table",
         table.properties().get("location"));
+    Assertions.assertEquals("MANAGED_TABLE", table.properties().get("table-type"));
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/trino/TrinoConnectorIT.java:400]: why there are some properties code?

Comment (yuqi1129) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/trino/TrinoConnectorIT.java]: I'm having trouble understanding your ideas. Can you provide more detail?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/HivePropertyMeta.java [hunk 2]
```diff
@@ -160,6 +220,22 @@ enum InsertExistingPartitionsBehavior {
...
+
+  public static SortingColumn sortingColumnFromString(String name) {
+    SortingColumn.Order order = ASCENDING;
+    String lower = name.toUpperCase(ENGLISH);
+    if (lower.endsWith(" ASC")) {
+      name = name.substring(0, name.length() - 4).trim();
+    } else if (lower.endsWith(" DESC")) {
+      name = name.substring(0, name.length() - 5).trim();
+      order = DESCENDING;
+    }
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/HivePropertyMeta.java:235]: Is it better to move these serialization functions to `SortingColumn`?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/HiveMetadataAdapter.java [hunk 3]
```diff
@@ -48,4 +79,147 @@ public Map<String, String> toGravitinoSchemaProperties(Map<String, Object> prope
...
+            ? (List<SortingColumn>) propertyMap.get(HIVE_SORT_ORDER_KEY)
+            : Collections.EMPTY_LIST;
+
+    if (!sortColumns.isEmpty()) {
+      Preconditions.checkArgument(
+          !bucketColumns.isEmpty() && bucketCount > 0,
+          "Bucket columns and bucket count must be specified when bucket count is set");
+    }
+
+    Map<String, String> properties = toGravitinoTableProperties(tableMetadata.getProperties());
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/HiveMetadataAdapter.java:113]: Yes, I think left it alone does no matter.

I will change this point.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/HivePropertyMeta.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuqi1129) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/HivePropertyMeta.java:235]: > Should we throw an error for spelling mistakes in ASC or DESC?

This method was copied from Hive connector.

[the remark this answers is not in the document]
Comment (yuqi1129) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/HivePropertyMeta.java:235]: > Is it better to move these serialization functions to `SortingColumn`?

OK

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/HiveMetadataAdapter.java [hunk 4]
```diff
@@ -48,4 +79,147 @@ public Map<String, String> toGravitinoSchemaProperties(Map<String, Object> prope
...
+        propertyMap.containsKey(HIVE_BUCKET_COUNT_KEY)
+            ? (int) propertyMap.get(HIVE_BUCKET_COUNT_KEY)
+            : 0;
+    List<SortingColumn> sortColumns =
+        propertyMap.containsKey(HIVE_SORT_ORDER_KEY)
+            ? (List<SortingColumn>) propertyMap.get(HIVE_SORT_ORDER_KEY)
+            : Collections.EMPTY_LIST;
+
+    if (!sortColumns.isEmpty()) {
+      Preconditions.checkArgument(
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/HiveMetadataAdapter.java]: > The use error need to use TrinoExecption with error code GRAVITINO_ILLEGAL_ARGUMENT

done. 

> How to handle errors on partition, sort and bucket?

I have no detailed idea about it, just let it as before.

**Review by diqiu50 — APPROVED** [review]
LGTM

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/SortingColumn.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import static java.util.Objects.requireNonNull;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.errorprone.annotations.Immutable;
+import io.trino.spi.connector.SortOrder;
+import java.util.Objects;
+
+@Immutable
+public class SortingColumn {
```
Comment (jerryshao) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/hive/SortingColumn.java:31]: If this is copied from somewhere, you should add reference here and update the LICENSE file.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/trino/TrinoConnectorIT.java [hunk 6]
```diff
@@ -739,6 +756,10 @@ void testHiveTableCreatedByGravitino() throws InterruptedException {
     Assertions.assertTrue(
         data.contains("input_format = 'org.apache.hadoop.hive.ql.io.orc.OrcInputFormat'"));
     Assertions.assertTrue(data.contains("serde_lib = 'org.apache.hadoop.hive.ql.io.orc.OrcSerde'"));
+    Assertions.assertTrue(data.contains("bucket_count = 4"));
+    Assertions.assertTrue(data.contains("bucketed_by = ARRAY['booleantype']"));
+    Assertions.assertTrue(data.contains("partitioned_by = ARRAY['binarytype']"));
+    Assertions.assertTrue(data.contains("sorted_by = ARRAY['longtype']"));
```
Comment (jerryshao) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/trino/TrinoConnectorIT.java:762]: I think you'd better add more test cases here, currently one is not enough to cover all the senarios.

Comment (yuqi1129) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/trino/TrinoConnectorIT.java:783]: I added more tests and tested the following case.
- Create partitioning, bucketing, and sorting by table by Trino and load it using Gravitino.
- Create partitioning, bucketing, and sorting by table by Gravitino and load it using Trino.

**Review by jerryshao — APPROVED** [review]

**Pull request merged**